core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "scale.hpp"
+
+#include <assert.h>
+#include <limits>
+#include <tuple>
+#include <type_traits>
+
+namespace kagome::common {
+
+  namespace scale {
+
+    namespace impl {
+      /**
+       * @brief Union class that helps representing integer value as an array
+       * @tparam T Integer type
+       */
+      template <class T>
+      union AsArray {
+        explicit AsArray(const T &t) : value_{t} {}
+
+        ///
+        uint8_t *begin() {
+          return static_cast<uint8_t *>(array_);
+        }
+
+        uint8_t *end() {
+          return begin() + size();
+        }
+
+        static const size_t size() {
+          return sizeof(T);
+        };
+
+        AsArray reversed() const {
+          AsArray r{value_};
+          for (auto left = r.begin(), right = r.end(); left < right;
+               ++left, --right) {
+            std::swap(*left, *right);
+          }
+          return r;
+        }
+
+        inline auto value() const {
+          return value_;
+        }
+
+        auto toLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+        auto fromLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+       private:
+        inline auto reverseIfHostNotLittleEndian() const {
+          if (__BYTE_ORDER == __LITTLE_ENDIAN) {
+            return value();
+          } else {
+            return reversed().value();
+          }
+        }
+
+        T value_;
+        uint8_t array_[sizeof(T)];
+      };
+
+      /**
+       * Helper class used for reinterpreting signed integers as unsigned and
+       * vice-versa
+       * @tparam T Signed or unsigned interger type
+       */
+      template <class T>
+      union SignReinterpreter {
+        using unsigned_type = typename std::make_unsigned<T>::type;
+        using signed_type = typename std::make_signed<T>::type;
+
+        SignReinterpreter<T>(unsigned_type uv) : unsigned_value_{uv} {}
+
+        SignReinterpreter<T>(signed_type sv) : signed_value_{sv} {}
+
+        unsigned_type asUnsigned() const {
+          return unsigned_value_;
+        }
+
+        signed_type asSigned() const {
+          return signed_value_;
+        }
+
+       private:
+        unsigned_type unsigned_value_;
+        signed_type signed_value_;
+      };
+
+      //      template <class T>
+      //      T encodeLittleEndian(T value) {
+      //        const int size = sizeof(T);
+      //        T result{0};
+      //        T multiplier = 1;
+      //        for (int i = 0; i < size; ++i) {
+      //          result += (value & 0xFF) * multiplier;
+      //          value >>= 8;
+      //          multiplier <<= 8;
+      //        }
+      //      }
```

Comment:
?

Reply:
Another way to encode little-endian any integer. Remove?

Reply:
Removed